// messaging-api-messenger/lib/Messenger.js
"use strict";Object.defineProperty(exports, "__esModule", { value: true });exports.default = void 0;

var _formData = _interopRequireDefault(require("form-data"));
var _invariant = _interopRequireDefault(require("invariant"));
var _isPlainObject = _interopRequireDefault(require("is-plain-object"));
var _lodash = _interopRequireDefault(require("lodash.omit"));
var _warning = _interopRequireDefault(require("warning"));function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}function ownKeys(object, enumerableOnly) {var keys = Object.keys(object);if (Object.getOwnPropertySymbols) {var symbols = Object.getOwnPropertySymbols(object);if (enumerableOnly) symbols = symbols.filter(function (sym) {return Object.getOwnPropertyDescriptor(object, sym).enumerable;});keys.push.apply(keys, symbols);}return keys;}function _objectSpread(target) {for (var i = 1; i < arguments.length; i++) {var source = arguments[i] != null ? arguments[i] : {};if (i % 2) {ownKeys(source, true).forEach(function (key) {_defineProperty(target, key, source[key]);});} else if (Object.getOwnPropertyDescriptors) {Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));} else {ownKeys(source).forEach(function (key) {Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));});}}return target;}function _defineProperty(obj, key, value) {if (key in obj) {Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true });} else {obj[key] = value;}return obj;}


















function validateQuickReplies(quickReplies) {
  // quick_replies is limited to 13
  (0, _invariant.default)(
  Array.isArray(quickReplies) && quickReplies.length <= 13,
  'quick_replies is an array and limited to 13');


  quickReplies.forEach(quickReply => {
    if (quickReply.content_type === 'text') {
      // title has a 20 character limit, after that it gets truncated
      (0, _invariant.default)(
      quickReply.title.trim().length <= 20,
      'title of quick reply has a 20 character limit, after that it gets truncated');


      // payload has a 1000 character limit
      (0, _invariant.default)(
      quickReply.payload.length <= 1000,
      'payload of quick reply has a 1000 character limit');

    }
  });
}

function createMessage(
msg,
options = {})
{
  const message = _objectSpread({},
  msg);


  if (
  options.quick_replies &&
  Array.isArray(options.quick_replies) &&
  options.quick_replies.length >= 1)
  {
    validateQuickReplies(options.quick_replies);
    message.quick_replies = options.quick_replies;
  }

  return message;
}

function createText(
text,
options = {})
{
  return createMessage({ text }, options);
}

function createMessageFormData(
payload,
filedata,
options = {})
{
  const message = _objectSpread({},
  payload);


  if (options.quick_replies) {
    validateQuickReplies(options.quick_replies);
    message.quick_replies = options.quick_replies;
  }

  const formdata = new _formData.default();

  formdata.append('message', JSON.stringify(message));
  formdata.append('filedata', filedata, (0, _lodash.default)(options, ['quick_replies']));

  return formdata;
}

function createAttachment(
attachment,
options = {})
{
  return createMessage(
  {
    attachment },

  options);

}

function createAttachmentFormData(attachment, filedata, options) {
  return createMessageFormData(
  {
    attachment },

  // $FlowFixMe
  filedata,
  options);

}

function createAudio(
audio,
options = {})
{
  const attachment = {
    type: 'audio',
    payload: {} };

  if (typeof audio === 'string') {
    attachment.payload.url = audio;
    return createAttachment(attachment, options);
  }
  if (audio && (0, _isPlainObject.default)(audio)) {
    attachment.payload = audio;
    return createAttachment(attachment, options);
  }

  return createAttachmentFormData(attachment, audio, options);
}

function createImage(
image,
options = {})
{
  const attachment = {
    type: 'image',
    payload: {} };

  if (typeof image === 'string') {
    attachment.payload.url = image;
    return createAttachment(attachment, options);
  }
  if (image && (0, _isPlainObject.default)(image)) {
    attachment.payload = image;
    return createAttachment(attachment, options);
  }

  return createAttachmentFormData(attachment, image, options);
}

function createVideo(
video,
options = {})
{
  const attachment = {
    type: 'video',
    payload: {} };

  if (typeof video === 'string') {
    attachment.payload.url = video;
    return createAttachment(attachment, options);
  }
  if (video && (0, _isPlainObject.default)(video)) {
    attachment.payload = video;
    return createAttachment(attachment, options);
  }

  return createAttachmentFormData(attachment, video, options);
}

function createFile(
file,
options = {})
{
  const attachment = {
    type: 'file',
    payload: {} };

  if (typeof file === 'string') {
    attachment.payload.url = file;
    return createAttachment(attachment, options);
  }
  if (file && (0, _isPlainObject.default)(file)) {
    attachment.payload = file;
    return createAttachment(attachment, options);
  }

  return createAttachmentFormData(attachment, file, options);
}

function createTemplate(
payload,
options = {})
{
  return createAttachment(
  {
    type: 'template',
    payload },

  options);

}

function createButtonTemplate(
text,
buttons,
options = {})
{
  return createTemplate(
  {
    template_type: 'button',
    text,
    buttons },

  options);

}

function createGenericTemplate(
elements,
options =


{})
{
  return createTemplate(
  {
    template_type: 'generic',
    elements,
    image_aspect_ratio: options.image_aspect_ratio || 'horizontal' },

  options);

}

function createListTemplate(
elements,
buttons,
options =


{})
{
  return createTemplate(
  {
    template_type: 'list',
    elements,
    buttons,
    top_element_style: options.top_element_style || 'large' },

  options);

}

function createOpenGraphTemplate(
elements,
options = {})
{
  return createTemplate(
  {
    template_type: 'open_graph',
    elements },

  options);

}

function createMediaTemplate(
elements,
options = {})
{
  return createTemplate(
  {
    template_type: 'media',
    elements },

  options);

}

function createReceiptTemplate(
attrs,
options = {})
{
  return createTemplate(_objectSpread({

    template_type: 'receipt' },
  attrs),

  options);

}

function createAirlineBoardingPassTemplate(
attrs,
options = {})
{
  return createTemplate(_objectSpread({

    template_type: 'airline_boardingpass' },
  attrs),

  options);

}

function createAirlineCheckinTemplate(
attrs,
options = {})
{
  return createTemplate(_objectSpread({

    template_type: 'airline_checkin' },
  attrs),

  options);

}

function createAirlineItineraryTemplate(
attrs,
options = {})
{
  return createTemplate(_objectSpread({

    template_type: 'airline_itinerary' },
  attrs),

  options);

}

function createAirlineUpdateTemplate(
attrs,
options = {})
{
  return createTemplate(_objectSpread({

    template_type: 'airline_update' },
  attrs),

  options);

}

function deprecated(name, fn) {
  return (...args) => {
    (0, _warning.default)(
    false,
    `\`Messenger.${name}\` is deprecated. Use \`Messenger.${fn.name}\` instead.`);

    return fn(...args);
  };
}

const Messenger = {
  createMessage,
  createText,
  createAttachment,
  createAudio,
  createImage,
  createVideo,
  createFile,
  createTemplate,
  createButtonTemplate,
  createGenericTemplate,
  createListTemplate: deprecated('createListTemplate', createListTemplate),
  createOpenGraphTemplate: deprecated(
  'createOpenGraphTemplate',
  createOpenGraphTemplate),

  createMediaTemplate,
  createReceiptTemplate,
  createAirlineBoardingPassTemplate,
  createAirlineCheckinTemplate,
  createAirlineItineraryTemplate,
  createAirlineUpdateTemplate };var _default =


Messenger;exports.default = _default;
